File: py/search/string/BoyerMoore.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class BoyerMoore:
    def search(self, text: str, pattern: str) -> int:
        maxValue, minValue = ord(max(text)), ord(min(text))
        nextBadChar, nextGoodSuffix = self.__getNextBadChar(pattern, maxValue - minValue + 1, minValue), self.__getNextGoodSuffix(pattern)

        index = 0
        while index <= len(text) - len(pattern):
            patternIndex = len(pattern) - 1
            while text[index + patternIndex] == pattern[patternIndex]:
                patternIndex -= 1
                if patternIndex < 0:
                    return index
            index += max(patternIndex - nextBadChar[ord(text[index + patternIndex]) - minValue], nextGoodSuffix[patternIndex])
        return -1

    def __getNextBadChar(self, text: str, nextSize: int, minValue: int) -> List[int]:
        next = [0 for _ in range(nextSize)]
        for i, value in enumerate(text):
            index = ord(value) - minValue
            if index >= 0 and index < nextSize:
                next[index] = i
        return next

    def __getNextGoodSuffix(self, text: str) -> List[int]:
        aux = [0 for _ in range(len(text))]
        aux[-1] = len(text)

        start, end = 0, len(text) - 1
        for i in range(len(text) - 2, -1, -1):
            if i > end and aux[i - 1 - start] < i - end:
                aux[i] = aux[i - 1 - start]
                continue

            if i < end:
                end = i
            start = i
            while end >= 0 and text[end] == text[end - 1 - start]:
                end -= 1
                aux[i] = start - end

        next = [len(text) for _ in range(len(text))]
        for i in range(len(text) - 1, -1, -1):
            if aux[i] != i + 1:
                continue

            for j in range(len(text) - i):
                if next[j] == len(text):
                    next[j] = len(text) - i - 1

        for i in range(len(text) - 1):
            next[-1 - aux[i]] = len(text) - i - 1
        return next

File: c/search/string/BoyerMoore.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int *getNextBadChar(char *text, int length, int nextSize, int minValue);
static int *getNextGoodSuffix(char *text, int length);
static int max(int i, int j);

int boyerMoore(char *text, char *pattern) {
    int length = strlen(text);
    int patternLength = strlen(pattern);

    int maxValue = text[0];
    int minValue = text[0];
    for (char *p = text + 1; '\0' != *p; ++p) {
        if (maxValue < *p) {
            maxValue = *p;
        }
        if (minValue > *p) {
            minValue = *p;
        }
    }
    int *nextBadChar = getNextBadChar(pattern, patternLength, maxValue - minValue + 1, minValue);
    int *nextGoodSuffix = getNextGoodSuffix(pattern, patternLength);

    int index = 0;
    while (index <= length - patternLength) {
        int patternIndex = patternLength - 1;
        while (text[index + patternIndex] == pattern[patternIndex]) {
            --patternIndex;
            if (patternIndex < 0) {
                return index;
            }
        }
        index += max(patternIndex - nextBadChar[text[index + patternIndex] - minValue], nextGoodSuffix[patternIndex]);
    }
    return -1;
}

static int *getNextBadChar(char *text, int length, int nextSize, int minValue) {
    int *next = (int *) malloc(sizeof(int) * nextSize);
    memset(next, 0, sizeof(int) * nextSize);
    for (int i = 0; i < length; ++i) {
        int index = text[i] - minValue;
        if (index >= 0 && index < nextSize) {
            next[index] = i;
        }
    }
    return next;
}

static int *getNextGoodSuffix(char *text, int length) {
    int *aux = (int *) malloc(sizeof(int) * length);
    memset(aux, 0, sizeof(int) * length);
    aux[length - 1] = length;

    int start = 0;
    int end = length - 1;
    for (int i = length - 2; i >= 0; --i) {
        if (i > end && aux[i + length - 1 - start] < i - end) {
            aux[i] = aux[i + length - 1 - start];
            continue;
        }

        start = i;
        if (i < end) {
            end = i;
        }
        while (end >= 0 && text[end] == text[end + length - 1 - start]) {
            --end;
            aux[i] = start - end;
        }
    }

    int *next = (int *) malloc(sizeof(int) * length);
    for (int i = 0; i < length; ++i) {
        next[i] = length;
    }
    for (int i = length - 1; i >= 0; --i) {
        if (aux[i] != i + 1) {
            continue;
        }

        for (int j = 0; j < length - i; ++j) {
            if (next[j] == length) {
                next[j] = length - i - 1;
            }
        }
    }

    for (int i = 0; i < length - 1; ++i) {
        next[length - 1 - aux[i]] = length - i - 1;
    }
    return next;
}

static int max(int i, int j) {
    int aux = ~((i > j) ? 1 : 0) + 1;
    return (i & aux) | (j & ~aux);
}
